contracts/FungibleTokenSwitchboard.cdc

```diff
@@ -0,0 +1,143 @@
+        /// receiverCapabilities
+        /// Dictionary holding the fungible token receiver capabilities, 
+        /// indexed by the fungible token vault type
+        ///
+        access(contract) var receiverCapabilities: {Type: Capability<&{FungibleToken.Receiver}>}
+
+        /// addNewVault adds a new fungible token receiver capability
+        ///                    to the switchboard resource
+        ///            
+        /// 
+        /// Parameters: capability: The capability to expose a certain fungible
+        /// token vault deposit function through {FungibleToken.Receiver} that
+        /// will be added to the switchboard
+        ///
+        pub fun addNewVault(capability: Capability<&{FungibleToken.Receiver}>) {
+            // Borrow a reference to the vault pointed by the capability we want
+            // to store inside the switchboard
+            let vaultRef = capability.borrow() 
+                ?? panic ("Cannot borrow reference to vault from capability")
+            // Use the vault reference type as key for storing the capability
+            self.receiverCapabilities[vaultRef.getType()] = capability
+            // Emit the event that indicates that a new capability has been added
+            emit VaultCapabilityAdded(type: vaultRef.getType())
+        }
+
+        /// removeVault removes a fungible token receiver capability 
+        ///                       from the switchboard resource
+        /// 
+        /// Parameters: capability: The capability to a fungible token vault 
+        ///                         to be removed from the switchboard
+        ///
+        pub fun removeVault(capability: Capability<&{FungibleToken.Receiver}>) {
+            // Borrow a reference to the vault pointed by the capability we want
+            // store inside the switchboard            
+            let vaultRef = capability.borrow() 
+                ?? panic ("Cannot borrow reference to vault from capability")
+            // Use the vault reference to find the capability to remove
+            self.receiverCapabilities.remove(key: vaultRef.getType())
+            // Emit the event that indicates that a new capability has been removed
+            emit VaultCapabilityRemoved(type: vaultRef.getType())            
+        }
+        
+        /// deposit Takes a fungible token vault and routes it to the proper
+        ///         fungible token receiver capability for depositing it
+        /// 
+        /// Parameters: from: The deposited fungible token vault resource
+        ///        
+        pub fun deposit(from: @FungibleToken.Vault) {
+            let depositedVaultCapability = self.receiverCapabilities[from.getType()] ?? 
+                panic ("The deposited vault is not available on this switchboard")
+            let vaultRef = depositedVaultCapability.borrow() ?? 
+                panic ("Can not borrow a reference to the the vault")
+            vaultRef.deposit(from: <- from)
+        }
```

Comment:
hmm to discuss:

- should we `panic` in deposit ?

It feels so wrong for me to panic here, though we will end up with resource otherwise.

Reply:
Any ideas of what we could do with the resource otherwise? We can't return it because that would change the deposit function signature

Reply:
how about we add a second function to public interface, something like ‘safeDeposit’? So if who is sending knows that we are switchboard, they can use safeDeposit (which returns NFT vault) 

this way we provide extended functionality depending on the client while preserving standard.

Reply:
That is an interesting idea. @alilloig Do you want to try adding that?

Reply:
I'm not sure if I'm understanding it properly, are we talking about something like this?
```
        /// safeDeposit Takes a fungible token vault and returns a reference to 
        ///             to the proper Vault from the switchboard, allowing the 
        ///             depositor to use the reference for depositing the funds
        ///             
        /// Parameters: vaultType: The type of the ft vault that wants to be deposited
        ///
        /// Returns: A reference to the required fungible token vault
        pub fun safeDeposit(vaultType: Type): &FungibleToken.Vault {

        }

Reply:
Something like this: 

```
pub fun safeDeposit(from: @FungibleToken.Vault): @FungibleToken.Vault {
    if let depositedVaultCapability = self.receiverCapabilities[from.getType()] {
        if let vaultRef = depositedVaultCapability.borrow() {
             vaultRef.deposit(from: <- from.withdraw(amount: from.balance) )
        }
    }
    return <- from 
}
```

so if any reason it cannot deposit, it will return the Vault back

Reply:
oh okey great!! I was not seeing the whole picture, I believe the point is to give payers a way to use the switchboard, probably focussed on the paying royalties use case, a way to have the transaction completed even if any of the receiver's vault is no available, neat! I actually love it thank you @bluesign !!!

Reply:
check the commit and let me know what you think guys. I coded it with code readability in mind, let me know if you find it "over commented" but I wanted to emphasize and make cristal clear the difference between that and the regular deposit function